src/cfnlint/rules/resources/codepipeline/CodepipelineStageActions.py

```diff
@@ -0,0 +1,254 @@
+                    'OutputArtifactRange': 0,
+                },
+            },
+            'Invoke': {
+                'Lambda': {
+                    'InputArtifactRange': (0, 5),
+                    'OutputArtifactRange': (0, 5),
+                }
+            }
+        },
+        'ThirdParty': {
+            'Source': {
+                'GitHub': {
+                    'InputArtifactRange': 0,
+                    'OutputArtifactRange': 1,
+                }
+            },
+        },
+    }
+
+    KEY_MAP = {
+        'InputArtifacts': 'InputArtifactRange',
+        'OutputArtifacts': 'OutputArtifactRange',
+    }
+
+    def check_artifact_counts(self, action, artifact_type, path):
+        """Check that artifact counts are within valid ranges."""
+        matches = []
+
+        action_type_id = action.get('ActionTypeId')
+        owner = action_type_id.get('Owner')
+        category = action_type_id.get('Category')
+        provider = action_type_id.get('Provider')
+
+        constraints = self.CONSTRAINTS.get(owner, {}).get(category, {}).get(provider, {})
+        if not constraints:
+            return matches
+        artifact_count = len(action.get(artifact_type, []))
+
+        constraint_key = self.KEY_MAP[artifact_type]
+        if isinstance(constraints[constraint_key], tuple):
+            min_, max_ = constraints[constraint_key]
+            if not (min_ <= artifact_count <= max_):
+                message = (
+                    'Action "{action}" declares {number} {artifact_type} which is not in '
+                    'expected range [{a}, {b}].'
+                ).format(
+                    action=action['Name'],
+                    number=artifact_count,
+                    artifact_type=artifact_type,
+                    a=min_,
+                    b=max_
+                )
+                matches.append(RuleMatch(
+                    path + [artifact_type],
+                    message
+                ))
+        else:
+            if artifact_count != constraints[constraint_key]:
+                message = (
+                    'Action "{action}" declares {number} {artifact_type} which is not the '
+                    'expected number [{a}].'
+                ).format(
+                    action=action['Name'],
+                    number=artifact_count,
+                    artifact_type=artifact_type,
+                    a=constraints[constraint_key]
+                )
+                matches.append(RuleMatch(
+                    path + [artifact_type],
+                    message
+                ))
+
+        return matches
+
+    def check_action_keys(self, action, path):
+        """Check that action has required keys inside."""
+        matches = []
+
+        if 'Name' not in action:
+            message = 'Expecting Name key in Action.'
+            matches.append(RuleMatch(
+                path,
+                message
+            ))
+        if 'ActionTypeId' not in action:
```

Comment:
Ummm this should be the same and covered by E3003 but its not working which means I think you found another bug.

```
"ActionTypeId": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-codepipeline-pipeline-stages-actions.html#cfn-codepipeline-pipeline-stages-actions-actiontypeid",
          "Required": true,
          "Type": "ActionTypeId",
          "UpdateType": "Mutable"
        },
```

Reply:
I apologize.  This was already fixed in the main branch.